File: makesite.py
import os
import random
import shutil
import string

from bs4 import BeautifulSoup
from jinja2 import Environment, FileSystemLoader
from markdown import markdown


def random_string(length=12):
    chars = string.ascii_lowercase + string.digits  # a-z and 0-9
    return "".join(random.choices(chars, k=length))


def immediate_subdirs(root_dir):
    return [name for name in os.listdir(root_dir) if os.path.isdir(os.path.join(root_dir, name))]


def copy_assets():
    copied_dirs = []
    for dir in immediate_subdirs("content"):
        random_id = random_string()
        # Copy assets from the content directory to the build directory
        src_dir = os.path.join("content", dir)
        dst_dir = os.path.join("build", random_id)
        os.makedirs(dst_dir)
        for filename in os.listdir(src_dir):
            src_file = os.path.join(src_dir, filename)
            dst_file = os.path.join(dst_dir, filename)
            if os.path.isfile(src_file):
                shutil.copy2(src_file, dst_file)  # copy2 preserves metadata
        copied_dirs.append((dir, random_id))
    return copied_dirs


def make_page(page_template, html_content):
    # Extract h1
    soup = BeautifulSoup(html_content, "html.parser")

    # Find all <h1> elements
    h1_tags = soup.find_all("h1")

    assert len(h1_tags) < 2, "More than one <h1> found in the content"
    assert len(h1_tags) == 1, "No <h1> found in the content"
    title = h1_tags[0].text.strip()

    output = page_template.render(title=title, content=html_content)
    return output, title


def make_pages(page_template):
    pages = [f[:-3] for f in os.listdir("content") if f.endswith(".md")]

    generated_pages = []
    for id in pages:
        with open(f"content/{id}.md", "r", encoding="utf-8") as f:
            md_content = f.read()
        page_html_content = markdown(md_content)
        output, title = make_page(page_template, page_html_content)
        random_id = random_string()
        with open(f"build/{random_id}.html", "w", encoding="utf-8") as f:
            f.write(output)
        generated_pages.append((id, random_id, title))
    return generated_pages


def make_homepage(jinja2_env, page_template, generated_pages):
    home_template = jinja2_env.get_template("home.html")
    home_html = home_template.render(pages=[(id, title) for id, _, title in generated_pages])
    output, _ = make_page(page_template, home_html)
    random_id = random_string()
    with open(f"build/{random_id}.html", "w", encoding="utf-8") as f:
        f.write(output)
    return random_id


def make_htaccess(generated_pages, copied_dirs, homepage_id):
    with open("build/.htaccess", "w", encoding="utf-8") as f:
        f.write(f"DirectoryIndex {homepage_id}.html\n")
        f.write("\n")
        f.write("RewriteEngine On\n")
        for id, secret_id, _ in generated_pages:
            f.write(f"RewriteRule ^{id}$ /{secret_id}.html [L]\n")
        for dir, random_id in copied_dirs:
            f.write(f"RewriteRule ^{dir}/(.*)$ /{random_id}/$1 [L]\n")


def run():
    assert not os.path.exists("build"), "Build directory already exists. Please remove it first."
    env = Environment(loader=FileSystemLoader("templates"))
    page_template = env.get_template("page.html")
    os.makedirs("build")
    generated_pages = make_pages(page_template)
    homepage_id = make_homepage(env, page_template, generated_pages)
    copied_dirs = copy_assets()
    make_htaccess(generated_pages, copied_dirs, homepage_id)


if __name__ == "__main__":
    run()

File: src/makesite.ts
import * as fs from 'fs';
import * as path from 'path';

import * as cheerio from 'cheerio';
import fsExtra from 'fs-extra';
import MarkdownIt from 'markdown-it';
import mathjax3 from 'markdown-it-mathjax3';
import pug from 'pug';

const markdown = new MarkdownIt({
  html: true,
  linkify: true
}).use(mathjax3);

/**
 * Generate a random string of specified length
 */
function randomString(length = 12): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let result = '';
  for (let i = 0; i < length; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

/**
 * Get immediate subdirectories of a directory
 */
function immediateSubdirs(rootDir: string): string[] {
  return fs.readdirSync(rootDir)
    .filter(name => fs.statSync(path.join(rootDir, name)).isDirectory());
}

/**
 * Copy assets from content directory to build directory
 */
function copyAssets(): [string, string][] {
  const copiedDirs: [string, string][] = [];
  for (const dir of immediateSubdirs('content')) {
    const randomId = randomString();
    const srcDir = path.join('content', dir);
    const dstDir = path.join('build', randomId);
    fs.mkdirSync(dstDir, { recursive: true });
    for (const filename of fs.readdirSync(srcDir)) {
      const srcFile = path.join(srcDir, filename);
      const dstFile = path.join(dstDir, filename);
      if (fs.statSync(srcFile).isFile()) {
        fsExtra.copySync(srcFile, dstFile);
      }
    }
    copiedDirs.push([dir, randomId]);
  }
  return copiedDirs;
}

/**
 * Create an HTML page using the page template and content
 */
function makePage(pageTemplate: pug.compileTemplate, htmlContent: string): [string, string] {
  const $ = cheerio.load(htmlContent);
  const h1Tags = $('h1');
  if (h1Tags.length > 1) {
    throw new Error('More than one <h1> found in the content');
  }
  if (h1Tags.length === 0) {
    throw new Error('No <h1> found in the content');
  }
  const title = h1Tags.first().text().trim();
  const output = pageTemplate({ title, content: htmlContent });

  return [output, title];
}

/**
 * Generate pages from markdown files
 */
function makePages(pageTemplate: pug.compileTemplate): [string, string, string][] {
  const pages = fs.readdirSync('content')
    .filter(f => f.endsWith('.md'))
    .map(f => f.slice(0, -3));
  const generatedPages: [string, string, string][] = [];
  for (const id of pages) {
    const mdContent = fs.readFileSync(`content/${id}.md`, 'utf-8');
    const pageHtmlContent = markdown.render(mdContent);
    const [output, title] = makePage(pageTemplate, pageHtmlContent);
    const secretId = randomString();
    fs.writeFileSync(`build/${secretId}.html`, output);
    generatedPages.push([id, secretId, title]);
  }
  return generatedPages;
}

/**
 * Generate the homepage
 */
function makeHomepage(
  homeTemplate: pug.compileTemplate,
  pageTemplate: pug.compileTemplate,
  generatedPages: [string, string, string][]
): string {
  const pages = generatedPages.map(([id, , title]) => [id, title] as [string, string]);
  const homeHtml = homeTemplate({ pages });
  const [output] = makePage(pageTemplate, homeHtml);
  const randomId = randomString();
  fs.writeFileSync(`build/${randomId}.html`, output);
  return randomId;
}

/**
 * Create .htaccess file for URL rewriting
 */
function makeHtaccess(
  generatedPages: [string, string, string][],
  copiedDirs: [string, string][],
  homepageId: string
): void {
  let htaccessContent = `DirectoryIndex ${homepageId}.html\n\n`;
  htaccessContent += 'RewriteEngine On\n';
  for (const [id, secretId] of generatedPages.map(([id, secretId]) => [id, secretId])) {
    htaccessContent += `RewriteRule ^${id}$ /${secretId}.html [L]\n`;
  }
  for (const [id] of generatedPages.map(([id]) => [id])) {
    htaccessContent += `RewriteRule ^${id}/$ /${id} [R=301,L]\n`;
  }
  for (const [dir, randomId] of copiedDirs) {
    htaccessContent += `RewriteRule ^${dir}/(.*)$ /${randomId}/$1 [L]\n`;
  }
  fs.writeFileSync('build/.htaccess', htaccessContent);
}

/**
 * Main function to run the site generator
 */
export function run(): void {
  if (!fs.existsSync('build')) {
    fs.mkdirSync('build');
  }
  if (fs.existsSync('build')) {
    for (const file of fs.readdirSync('build')) {
      const filePath = path.join('build', file);
      fsExtra.removeSync(filePath);
    }
  }
  const pageTemplate = pug.compileFile('templates/page.pug');
  const homeTemplate = pug.compileFile('templates/home.pug');
  const generatedPages = makePages(pageTemplate);
  const homepageId = makeHomepage(homeTemplate, pageTemplate, generatedPages);
  const copiedDirs = copyAssets();
  makeHtaccess(generatedPages, copiedDirs, homepageId);
  console.log('Site generated successfully!');
}

run();
